### ISSUE TITLE
questions about run_scaled_dot_product_attention.

### ISSUE THREAD

**Issue by cbacode** [stanford-cs336/assignment1-basics#19]
Hello there, I am appreciated for your excellent class and I felt sorry for my poor English. However, I encountered a problem while trying to implement the function `run_scaled_dot_product_attention()`. The original type hint for mask is `Float[Tensor, " ... queries keys"]`, but I found that the tensor mask provided by the test function is actually `Bool[Tensor, " ... queries keys"]`, which could make a difference in the code. The fixed code may be as follows:
```python
def run_scaled_dot_product_attention(
    Q: Float[Tensor, " ... queries d_k"],
    K: Float[Tensor, " ... keys d_k"],
    V: Float[Tensor, " ... values d_v"],
    mask: Bool[Tensor, " ... queries keys"] | None = None,
) -> Float[Tensor, " ... queries d_v"]:
```

Besides, I found it difficult implementing run_rope because the fact that the power of theta in the pdf is an odd number, but I only passed the test until I change it into an even number. So I think the expression in the pdf may not approprate. Here is my full implementation, and the line `exp = float(2 * k) / float(d_k)` is the most important.
<img width="1141" height="174" alt="Image" src="https://github.com/user-attachments/assets/0ced4165-1db3-48d3-93e2-69dd200cae9d" />

```python
from einops import einsum, rearrange
from jaxtyping import Float, Int
import torch
from torch import nn
from torch import Tensor
import math

# uv run pytest -k run_rope

def run_rope(
    d_k: int,
    theta: float,
    max_seq_len: int,
    in_query_or_key: Float[Tensor, " ... sequence_length d_k"],
    token_positions: Int[Tensor, " ... sequence_length"],
) -> Float[Tensor, " ... sequence_length d_k"]:
    """
    Run RoPE for a given input tensor.

    Args:
        d_k (int): Embedding dimension size for the query or key tensor.
        theta (float): RoPE parameter.
        max_seq_len (int): Maximum sequence length to pre-cache if your implementation does that.
        in_query_or_key (Float[Tensor, "... sequence_length d_k"]): Input tensor to run RoPE on.
        token_positions (Int[Tensor, "... sequence_length"]): Tensor of shape (batch_size, sequence_length) with the token positions
    Returns:
        Float[Tensor, " ... sequence_length d_k"]: Tensor with RoPEd input.
    """
    l = RoPE(theta, d_k, max_seq_len, in_query_or_key.device)
    return l.forward(in_query_or_key, token_positions)

def singleton(cls):
    instances = {}
    def get_instance(*args, **kwargs):
        if cls not in instances:
            instances[cls] = cls(*args, **kwargs)
        return instances[cls]
    return get_instance

@singleton
class RoPE(nn.Module):
    def __init__(self, theta: float, d_k: int, max_seq_len: int, device=None):
        """ 
        Construct the RoPE module and create buffers if needed.
            theta: float Θ value for the RoPE
            d_k: int dimension of query and key vectors
            max_seq_len: int Maximum sequence length that will be inputted
            device: torch.device | None = None Device to store the buffer on
        """
        if not hasattr(self, 'cache'):
            super().__init__()
            self.theta = theta
            self.d_k = d_k
            self.max_seq_len = max_seq_len
            self.device = device
            # Compute and save cache
            cache = []
            for i in range(max_seq_len):
                inside_cache = []
                for k in range(d_k // 2):
                    exp = float(2 * k) / float(d_k)
                    angle = float(i) / pow(theta, exp)
                    arr = torch.tensor([[math.cos(angle), math.sin(angle)], [-math.sin(angle), math.cos(angle)]], device = device)
                    inside_cache.append(arr)
                cache.append(torch.stack(inside_cache))
            cache = torch.stack(cache)
            
            self.register_buffer("cache", cache, persistent=False)
        return
    
    def forward(self, x: torch.Tensor, token_positions: torch.Tensor) -> torch.Tensor:
        """ 
        Process an input tensor of shape (..., seq_len, d_k) and return a tensor of the same shape. Note that you should tolerate x with an arbitrary number of batch dimensions. You should assume that the token positions are a tensor of shape (..., seq_len) specifying the token positions of x along the sequence dimension.
        You should use the token positions to slice your (possibly precomputed) cos and sin tensors along the sequence dimension
        """
        d_k = x.size(-1)

        # rot is a (..., seq_len, d_k / 2, 2, 2) Tensor
        # if rot is a (..., seq_len, d_k, d_k) Tensor, then we can directly multiply it.
        rot = self.cache[token_positions]
        
        # Change x into a (..., seq_len, d_k, 1) Tensor
        # Break x into (..., seq_len, 2) Tensors and see x as a (..., seq_len, 2, d_k / 2) Tensor, for each Tensor in last dimension, manully rotate it and finally stack them together.
        x = rearrange(x, " ... seq_len d_k -> d_k ... seq_len")
        rot = rearrange(rot, "... seq_len d_k_2 row col -> d_k_2 ... seq_len row col")

        res = []
        for i in range(d_k // 2):
            left = torch.stack([x[i * 2], x[i * 2 + 1]], -1)
            res.append(einsum(left, rot[i], "... seq_len col, ... seq_len col row -> ... seq_len row"))
        return torch.cat(res, -1)
    
    def brute_forward(self, x: torch.Tensor, token_positions: torch.Tensor) -> torch.Tensor:
        d_k = x.size(-1)
        # Compute and save cache
        cache = []
        for i in range(self.max_seq_len):
            inside_cache = torch.tensor([[0.0] * d_k] * d_k)
            for k in range(d_k // 2):
                exp = float(2 * k) / float(d_k)
                angle = float(i) / pow(self.theta, exp)
                inside_cache[2 * k][2 * k] = math.cos(angle)
                inside_cache[2 * k][2 * k + 1] = math.sin(angle)
                inside_cache[2 * k + 1][2 * k] = -math.sin(angle)
                inside_cache[2 * k + 1][2 * k + 1] = math.cos(angle)
            cache.append(inside_cache)
        cache = torch.stack(cache)
        
        rot = cache[token_positions]
        return einsum(x, rot, "... seq_len d_k_1, ... seq_len d_k_1 d_k -> ... seq_len d_k")
```

**Comment by marcelroed** [comment]
Hey, sorry for the delay. The typing issue is definitely right and should be fixed, but I thought the RoPE issue was fixed in the last update. I will re-review soon and get back to you.

**Comment by cbacode** [comment]
Thank you for your reply! I see the last isuue and I tried to explain the RoPE issue again 12 hours ago, and I tried to use "odd numbers" and "even numbers" in the first comment to show that the RoPE issue was fixed in the wrong way. That is the reason why I mentioned RoPE in this issue. Your class is definately awesome and I have received multiply recommandations for your class. I sincerely hope this repo can be better!

**Comment by cbacode** [comment]
Sorry for my grammar mistakes and I have fixed them up.

**Comment by marcelroed** [comment]
Sorry about the RoPE issue, it should be fixed once I push the write-up update detailed in #22! Meanwhile, I also fixed the typing in 904e9ee. Let me know if this looks good to you.

**Comment by cbacode** [comment]
The typing in function `run_scaled_dot_product_attention` looks perfect! And it could be even better if we can be reminded that if we create the mask like:
```python
t = torch.ones(sequence_length, sequence_length, device = self.device) 
mask = torch.triu(t).to(torch.bool).transpose(1, 0)
```
then `torch.triu` will return a tensor in type `Float`, and we might use transpose to get the correct mask in `Problem (multihead_self_attention)`
Besides, don't forget to checkout the typing issue in `run_load_checkpoint`, sincerely appreciate your hard work!

**Comment by marcelroed** [comment]
Thanks! I would actually recommend not to use triu. Instead you can use a comparison between indices along two axes, for instance:

```python3
qi = torch.arange(n_queries)
kj = torch.arange(n_keys)

mask = qi[:, None] >= kj[None, :]
```

I'll update the typing for the `run_load_checkpoint` function.
